### PR TITLE
add support for field name convert. e.g. order_id <-> orderId

### PR DESCRIPTION
**Pull request by rolandhe**
## What is the purpose of the change

Adapt non-camelcase property name to camelcase. In a gateway application the data from legacy application such as php is non-camelcase style, but in general java use camelcase. This is complex to convert them in user application, it is necessary to provide the function in Dubbo.

## Brief changelog

Add GenericAlias annotation.
change PojoUtils class, change the realize and generalize method.

## Verifying this change

Follow this checklist to help us incorporate your contribution quickly and easily:

- [x] Make sure there is a [GITHUB_issue](https://github.com/apache/dubbo/issues/4286) field for the change (usually before you start working on it). Trivial changes like typos do not require a GITHUB issue. Your pull request should address just this issue, without pulling in other changes - one PR resolves one issue.
- [ ] Format the pull request title like `[Dubbo-XXX] Fix UnknownException when host config not exist #XXX`. Each commit in the pull request should have a meaningful subject line and body.
- [ ] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
- [ ] Write necessary unit-test to verify your logic correction, more mock a little better when cross module dependency exist. If the new feature or significant change is committed, please remember to add sample in [dubbo samples](https://github.com/apache/dubbo-samples) project.
- [ ] Run `mvn clean install -DskipTests=false` & `mvn clean test-compile failsafe:integration-test` to make sure unit-test and integration-test pass.
- [ ] If this contribution is large, please follow the [Software Donation Guide](https://github.com/apache/dubbo/wiki/Software-donation-guide).

### CODE REVIEW

File: dubbo-common/src/main/java/org/apache/dubbo/common/annotations/GenericAlias.java [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+
+/**
+ * Indicate alias for field when generic invoking. <br>
+ * <p>
+ *     sometimes you receive json string which contains non-camelcase key name,e.g. order_id,
+ *     but the java bean needs camelcase name,e.g orderId, you should convert manually each other,
+ *     basing GenericAlias annotation it could be converted automatically.
+ * </p>
+ *
+ * @author rolandhe
```
Comment (ralf0131) [inline dubbo-common/src/main/java/org/apache/dubbo/common/annotations/GenericAlias.java]: Hi, please remove the author name.

Comment (rolandhe) [reply dubbo-common/src/main/java/org/apache/dubbo/common/annotations/GenericAlias.java]: done.tks.

**Comment by beiwei30** [comment]
@rolandhe, this is very interesting feature, would you mind to add some unit test once this code gets merged?

**Comment by rolandhe** [comment]
@beiwei30 I covered realize, I changed getField method to support GenericAlias, It could be removed when you resolve the conflict. I modify the getField way of calling cls.getDeclaredField(fieldName) then loop, but only loop and get GenericAlias annotation. And I will fix and commit a demo tomorrow.

**Comment by rolandhe** [comment]
@beiwei30 I recommit code and add a demo for GenericAlias, pls check again.

**Comment by rolandhe** [comment]
@beiwei30 I found there are some conflicts, but I can't resolve, pls resolve them, tks.

**Comment by rolandhe** [comment]
@cvictory  I changed GenericAlias annotation to GenericFeature, and add more features, e.g. ignore  field, number to string.

File: dubbo-common/src/main/java/org/apache/dubbo/common/utils/PojoUtils.java [hunk 2]
```diff
@@ -1,628 +1,1012 @@
...
+        }
+        return clazz;
+    }
+
+    private static Object newInstance(Class<?> cls) {
+        try {
+            return cls.newInstance();
+        } catch (Throwable t) {
+            try {
+                Constructor<?>[] constructors = cls.getDeclar
```
Comment (lkj41110) [inline dubbo-common/src/main/java/org/apache/dubbo/common/utils/PojoUtils.java]: Is it better to have a cache for this method?And i think it is better in `ReflectUtils.class`

**Comment by chickenlj** [comment]
Please, check the comments and resolve the issues pointed.

**Comment by chickenlj** [comment]
Let's temporarily postpone this to 2.7.7, if the author responds just in time we can move it back to 2.7.6

File: dubbo-common/src/main/java/org/apache/dubbo/common/utils/PojoUtils.java [hunk 3]
```diff
@@ -1,628 +1,1012 @@
...
-    }
-
-    public static boolean isPojo(Class<?> cls) {
-        return !ReflectUtils.isPrimitives(cls)
-                && !Collection.class.isAssignableFrom(cls)
-                && !Map.class.isAssignableFrom(cls);
-    }
-
-}
+/*
```
Comment (AlbumenJ) [inline dubbo-common/src/main/java/org/apache/dubbo/common/utils/PojoUtils.java:1]: please recover file end-line format

File: dubbo-demo/dubbo-generic-demo/dubbo-generic-demo-consumer/src/main/java/org/apache/dubbo/demo/consumer/Application.java [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+    /**
+     * In order to make sure multicast registry works, need to specify '-Djava.net.preferIPv4Stack=true' before
+     * launch the application
+     */
+    public static void main(String[] args) {
+        ReferenceConfig<GenericService> reference = new ReferenceConfig<GenericService>();
+        // 弱类型接口名
+        reference.setInterface("org.apache.dubbo.demo.provider.DemoGenericService");
+        reference.setVersion("1.0.0");
+        // 声明为泛化接口
```
Comment (AlbumenJ) [inline dubbo-demo/dubbo-generic-demo/dubbo-generic-demo-consumer/src/main/java/org/apache/dubbo/demo/consumer/Application.java:35]: pls comment in English

File: dubbo-demo/dubbo-generic-demo/dubbo-generic-demo-provider/src/main/java/org/apache/dubbo/demo/provider/RequestDemo.java [hunk 5]
```diff
@@ -0,0 +1,55 @@
+package org.apache.dubbo.demo.provider;
+
+import org.apache.dubbo.common.annotations.GenericFeature;
+
+import java.util.Map;
+
+/**
+ * @author hexiufeng
```
Comment (AlbumenJ) [inline dubbo-demo/dubbo-generic-demo/dubbo-generic-demo-provider/src/main/java/org/apache/dubbo/demo/provider/RequestDemo.java:8]: pls remove author field, thx.

**Review by AlbumenJ — CHANGES_REQUESTED** [review]

**Comment by AlbumenJ** [comment]
Close for long time no response.

Please feel free to reopen if you have any question.

If you think these changes still useful in the latest master branch, please submit a new pull request.

**Pull request closed without merging**